crates/vm/src/compute.rs

```diff
@@ -0,0 +1,141 @@
+///
+/// Computes as many programs as is the input to this op in parallel.
+/// Each compute program executes on a constructed VM that has
+/// - Read-only access to parent VM memory
+/// - Parent VM stack with an additional value on top that is the compute index
+/// - Parent program counter + 1
+///
+/// When compute program returns after seeing op ComputeEnd or when ops come to an end,
+/// parent VM memory is updated to be the concatenation of children's memories.
+///
+/// Intended usage is parallelizable computation that can be distributed to self index-aware compute programs.
+///
+/// Returns resulting program counter and spent gas.
+pub fn compute<S, OA, OG>(inputs: ComputeInputs<S, OA, OG>) -> OpResult<(usize, Gas), S::Error>
+where
+    S: StateReads,
+    OA: OpAccess<Op = Op>,
+    OA::Error: Into<OpError<S::Error>>,
+    OG: OpGasCost,
+{
+    let ComputeInputs {
+        mut pc,
+        stack,
+        memory,
+        mut parent_memory,
+        repeat,
+        cache,
+        access,
+        state_reads,
+        op_access,
+        op_gas_cost,
+        gas_limit,
+    } = inputs;
+
+    // Pop the number of compute threads to spawn.
+    let compute_breadth = stack
+        .pop()
+        .map_err(|e| OpError::Compute(ComputeError::Stack(e)))?;
+    if compute_breadth < 1 {
+        return Err(OpError::Compute(ComputeError::<S::Error>::InvalidBreadth(
+            compute_breadth,
+        )));
+    }
+
+    // Append parent memory to be read by spawned threads.
+    if parent_memory.len() < MAX_COMPUTE_DEPTH {
+        parent_memory.push(Arc::new(memory.to_owned()));
+    } else {
+        return Err(ComputeError::DepthReached(MAX_COMPUTE_DEPTH).into());
+    }
+
+    // Compute in parallel.
+    let results: Result<Vec<(Gas, usize, Memory)>, _> = (0..compute_breadth)
+        .into_par_iter()
+        .map(|compute_index| {
+            // Clone stack and push compute program index.
+            let mut stack = stack.clone();
+            stack
+                .push(compute_index)
+                .map_err(|e| ExecError(pc, OpError::Compute(e.into())))?;
+
+            let mut vm = Vm {
+                pc: pc + 1,
+                stack,
+                memory: Memory::new(),
+                parent_memory: parent_memory.clone(),
+                repeat: repeat.clone(),
+                cache: cache.clone(),
+            };
+
+            // Execute child VM.
+            vm.exec(
+                access.clone(),
+                state_reads,
+                op_access.clone(),
+                op_gas_cost,
+                gas_limit,
+            )
+            .map(|gas| (gas, vm.pc, vm.memory))
+        })
+        .collect();
+
+    let oks = results.map_err(|e| OpError::Compute(ComputeError::Exec(Box::new(e))))?;
+
+    // Assign concatanated compute memories to parent memory.
+    // FIXME: avoid cloning the memory and extend the original memory
+    // in a more straightforward way than alloc + store_range
```

Comment:
Yeh I agree with this. Sum the sizes of the child memories, allocate that and then store range each one